students/anbaopham/lesson05/Assignment_3.py

```diff
@@ -0,0 +1,195 @@
+            l[1:] = [" ".join(l[1:])]
+        first_last_name_list.append(l)
+    return first_last_name_list
+
+def Thank_You_letter_template(first_name, last_name, donation_amount):
+    first_name = first_name
+    last_name = last_name
+    donation_amount = donation_amount
+
+    template = """Dear {first_name} {last_name},
+    Thank you for your very kind donation of ${donation_amount}
+    It will be put to very good use.
+        Sincerely,
+            -The Team"""
+
+    letter_data ={"first_name": first_name, "last_name": last_name, "donation_amount": donation_amount}
+
+    filename = f"{first_name}_{last_name}.txt"
+    with open(filename, "w") as fp:
+        fp.write(template.format(**letter_data))
+
+
+
+def send_ALL_Thank_You():
+    first_last_name_list= get_first_last_name()
+    full_name_list = get_name_list()
+    for i in range(len(first_last_name_list)):
+        first_name = first_last_name_list[i][0]
+        last_name = first_last_name_list[i][1]
+        Thank_You_letter_template(first_name, last_name, 10)
+
+
+
+def create_a_Report():
+    l = (["Donor Name", "Total Given", "Num Gifts", "Average Gift"])
+    temp_list = []
+    l1 = []
+
+    total_given =0
+    for x in donor_db:
+        temp_list.append(x[0])
+        tg = sum(list((x[1])))
+        total_given = str(tg)
+        temp_list.append(total_given)
+        ng = len(x[1])
+        num_gift = str(ng)
+        temp_list.append(num_gift)
+        #ave_gift = str(round(total_given/num_gift,2))
+        aG = round(tg/ng,2)
+        ave_gift = str(aG)
+        temp_list.append(ave_gift)
+        #report_table.append(x[0], total_given, num_gift, ave_gift)
+    m = 0
+    while m< len(temp_list):
+        l1.append(temp_list[m:m+4])
+        m = m+4
+
+    l1 =sorted(l1, key = sort_key)
+
+    l1.insert(0,l)
+
+    new_list = []
+    #for i in l1:
+    #    for j in i:
+    #        new_list.append(len(j))
+    #        l = max(new_list)
+
+    [new_list.append(len(j)) for i in l1 for j in i]
+    l = max(new_list)
+    space = " "*10
+    width = l
+    #print(l1)
+    for a,b,c,d in l1:
+        print(f'{a:<{width}}{space}{b:>{width}}{space}{c:>{width}}{space}{d:>{width}}')
+
+def exit_program():
+    print("Bye!")
+    sys.exit()  # exit the interactive script
+
+response_dict ={"1": send_a_Thank_You, "2": create_a_Report, "3": send_ALL_Thank_You, "4": exit_program}
+def main():
+    while True:
+        response = input(prompt)  # continuously collect user selection
+        # now redirect to feature functions based on the user selection
+        if response in response_dict:
+            response_dict.get(response)()
+        else:
+            print("Not a valid option!")
+
+        #try:
+        #    response_dict.get(response)()
+        #except:
+        #    print("Not a valide option!")
+
+
+
+
+if __name__ == "__main__":
+    # don't forget this block to guard against your code running automatically if this module is imported
+    main()
```

Comment:
Same issues with this version of mailroom as prior.  All functions should be lowercase with words separated by underscores. Report should be sorted by total given in descending order.  Also, when selecting report after adding donor, code is failing and stops running.

Reply:
I added sort_total_given function.  Somehow it doesn't sort as expected.  What should I do?

Did you mean that the code failed when you selected option 1 then select option 2?  I tested it and don't see error.  Please review it again or let me know what line cause the failure.  Thanks.